aggregator/election_mgr.go

```diff
@@ -0,0 +1,518 @@
+	campaignCreateErrors tally.Counter
+	campaignErrors       tally.Counter
+	leaderErrors         tally.Counter
+	leaderNotChanged     tally.Counter
+	resignErrors         tally.Counter
+	followerResign       tally.Counter
+	resignNotFollower    tally.Counter
+	followerToLeader     tally.Counter
+	leaderToFollower     tally.Counter
+	electionState        tally.Gauge
+}
+
+func newElectionManagerMetrics(scope tally.Scope) electionManagerMetrics {
+	return electionManagerMetrics{
+		campaignCreateErrors: scope.Counter("campaign-create-errors"),
+		campaignErrors:       scope.Counter("campaign-errors"),
+		leaderErrors:         scope.Counter("leader-errors"),
+		leaderNotChanged:     scope.Counter("leader-not-changed"),
+		resignErrors:         scope.Counter("resign-errors"),
+		resignNotFollower:    scope.Counter("resign-not-follower"),
+		followerResign:       scope.Counter("follower-resign"),
+		followerToLeader:     scope.Counter("follower-to-leader"),
+		leaderToFollower:     scope.Counter("leader-to-follower"),
+		electionState:        scope.Gauge("election-state"),
+	}
+}
+
+type electionManager struct {
+	sync.RWMutex
+
+	nowFn           clock.NowFn
+	logger          xlog.Logger
+	reportInterval  time.Duration
+	campaignOpts    services.CampaignOptions
+	electionOpts    services.ElectionOptions
+	campaignRetrier xretry.Retrier
+	changeRetrier   xretry.Retrier
+	electionKeyFmt  string
+	leaderService   services.LeaderService
+	leaderValue     string
+
+	state            electionManagerState
+	doneCh           chan struct{}
+	notifyCh         chan struct{}
+	resignWg         *sync.WaitGroup
+	electionKey      string
+	electionState    ElectionState
+	changeInProgress bool
+	changeCancelFn   context.CancelFunc
+	changeWg         sync.WaitGroup
+	sleepFn          sleepFn
+	metrics          electionManagerMetrics
+}
+
+// NewElectionManager creates a new election manager.
+func NewElectionManager(opts ElectionManagerOptions) ElectionManager {
+	instrumentOpts := opts.InstrumentOptions()
+	campaignOpts := opts.CampaignOptions()
+	campaignRetrier := xretry.NewRetrier(opts.CampaignRetryOptions().SetForever(true))
+	changeRetrier := xretry.NewRetrier(opts.ChangeRetryOptions().SetForever(true))
+	mgr := &electionManager{
+		nowFn:           opts.ClockOptions().NowFn(),
+		logger:          instrumentOpts.Logger(),
+		reportInterval:  instrumentOpts.ReportInterval(),
+		campaignOpts:    campaignOpts,
+		electionOpts:    opts.ElectionOptions(),
+		campaignRetrier: campaignRetrier,
+		changeRetrier:   changeRetrier,
+		electionKeyFmt:  opts.ElectionKeyFmt(),
+		leaderService:   opts.LeaderService(),
+		leaderValue:     campaignOpts.LeaderValue(),
+		state:           electionManagerNotOpen,
+		doneCh:          make(chan struct{}),
+		notifyCh:        make(chan struct{}, 1),
+		electionState:   FollowerState,
+		sleepFn:         time.Sleep,
+		metrics:         newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+	go mgr.reportMetrics()
```

Comment:
Shouldn't perhaps this be launched from `Open(...)`?  It seems the only way to break from the metrics loop is to read the close of the `doneCh` which can only be achieved if you've opened the manager.